src/webkit/wkPage.ts

```diff
@@ -384,14 +384,23 @@ export class WKPage implements PageDelegate {
     await this._forAllSessions(session => WKPage._setEmulateMedia(session, mediaType, colorScheme));
   }
 
-  async setViewport(viewport: types.Viewport): Promise<void> {
-    const width = viewport.width;
-    const height = viewport.height;
-    const fixedLayout = !!viewport.isMobile;
-    const deviceScaleFactor = viewport.deviceScaleFactor || 1;
-    this._page._state.hasTouch = !!viewport.isMobile;
+  async setViewportSize(viewportSize: types.Size): Promise<void> {
+    assert(this._page._state.viewportSize === viewportSize);
+    await this._updateViewport();
+  }
+
+  async _updateViewport(): Promise<void> {
+    let viewport = this._page.browserContext()._options.viewport || { width: 0, height: 0 };
+    const viewportSize = this._page._state.viewportSize;
+    if (viewportSize)
+      viewport = { ...viewport, ...viewportSize };
     await Promise.all([
-      this._pageProxySession.send('Emulation.setDeviceMetricsOverride', {width, height, fixedLayout, deviceScaleFactor }),
+      this._pageProxySession.send('Emulation.setDeviceMetricsOverride', {
+        width: viewport.width,
+        height: viewport.height,
+        fixedLayout: !!viewport.isMobile,
+        deviceScaleFactor: viewport.deviceScaleFactor || 1
+      }),
       this._updateState('Page.setTouchEmulationEnabled', { enabled: !!viewport.isMobile }),
```

Comment:
A little unfortunate we do it on setsize

Reply:
Avoided this with a flag :)